front/components/assistant_builder/AssistantBuilderTablesModal.tsx

```diff
@@ -55,45 +114,199 @@ export default function AssistantBuilderTablesModal({
+        setNodeIntializationError(true);
+        throw e;
+      })
+      .finally(() => {
+        setIsInitializingNodes(false);
+      });
+  }, [
+    selectedDataSource,
+    tablesQueryConfiguration,
+    isInitializingNodes,
+    selectedManagedTables.length,
+    initializeManagedTables,
+    nodeIntializationError,
+  ]);
+
+  useEffect(() => {
+    if (
+      !selectedDataSource ||
+      !selectedDataSource.connectorId ||
+      isInitializingParents ||
+      !!parentsIntializationError ||
+      !!Object.keys(parentsById).length ||
+      !selectedManagedTables.length
+    ) {
+      return;
+    }
+    setIsInitializingParents(true);
+    void initializeParents(
+      selectedManagedTables.map((n) => n.internalId),
+      selectedDataSource
+    )
+      .catch((e) => {
+        setParentsIntializationError(true);
+        throw e;
+      })
+      .finally(() => {
+        setIsInitializingParents(false);
+      });
+  }, [
+    selectedDataSource,
+    selectedManagedTables,
+    parentsById,
+    initializeParents,
+    isInitializingParents,
+    parentsIntializationError,
+  ]);
+
+  async function save() {
+    if (!selectedDataSource) {
+      return;
+    }
+    setIsSaving(true);
+
+    try {
+      const tableIds = removeNulls(
+        selectedManagedTables.map((n) => n.dustTableId)
+      );
+      const tables = await Promise.all(
+        tableIds.map(async (id) => {
+          const tableRes = await fetch(
+            `/api/w/${owner.sId}/data_sources/${selectedDataSource.name}/tables/${id}`
+          );
+          const { table } = (await tableRes.json()) as { table: CoreAPITable };
+          return table;
+        })
+      );
+      const configs = tables.map((table) => ({
+        workspaceId: owner.sId,
+        dataSourceId: table.data_source_id,
+        tableId: table.table_id,
+        tableName: table.name,
+      }));
+
+      onSave(configs);
+    } finally {
+      setIsSaving(false);
+    }
+  }
 
   const onClose = () => {
     setOpen(false);
     setTimeout(() => {
       setSelectedDataSource(null);
-      setSelectedTable(null);
+      setSelectedManagedTables([]);
+      setIsSaving(false);
+      setIsInitializingNodes(false);
+      setNodeIntializationError(false);
     }, 200);
   };
 
   return (
     <Modal
       isOpen={isOpen}
-      onClose={() => {
-        if (selectedDataSource !== null) {
-          setSelectedDataSource(null);
-        } else {
-          onClose();
-        }
-      }}
+      onClose={onClose}
+      isSaving={isSaving}
       onSave={() => {
-        if (selectedTable) {
-          onSave(selectedTable);
-        }
+        void save().then(onClose);
       }}
-      hasChanged={!!selectedTable}
+      hasChanged={!!selectedManagedTables.length}
       variant="full-screen"
       title="Select Tables"
     >
       <div className="w-full pt-12">
-        {!selectedDataSource ? (
+        {isInitializingNodes || isInitializingParents ? (
```

Comment:
Can we avoid nested nested ternaries here?

Reply:
Like creating another component ? 😬  
Will see how I can remove, but there are no nested ternaries here

Reply:
Indeed, it can probably be extracted in a top-level function/component.

Reply:
I'm kind of torn here, If I make another component then I have to drill down a ton of props, I find the result actually less readable than this (unfortunate) ternary